### PR TITLE
[#173338803] Integration test for activation phase

### PR DESCRIPTION
**Pull request by fabriziofff**
**Short description:**
This pr adds some integration tests to verify the behaviour of the activation flow with different backend messages.

**List of changes proposed in this pull request:**
- added `integration/activationIntegrationTests.ts`

**How to test:**
run `activationIntegrationTests.ts`.

### CODE REVIEW

File: ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/activationIntegrationTests.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+      bonusActivationSaga(startBonusActivation, getActivationById)
+    )
+      .withReducer(activationReducer)
+      .provide([
+        [
+          select(navigationCurrentRouteSelector),
+          fromNullable(BONUSVACANZE_ROUTES.ACTIVATION.LOADING)
+        ],
+        [
+          matchers.call.fn(startBonusActivation),
```
Comment (Undermaken) [inline ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/activationIntegrationTests.ts:60]: I found a bug by adding the case when the server returns something the app can't decode
```typescript
[
   matchers.call.fn(startBonusActivation),
   left([{ context: [], value: new Error("decoding failure") }])
]
```

I guess this scenario is also valuable.
The fix https://github.com/pagopa/io-app/pull/1931/commits/8568e5416fb0474971c56d54fa00dc93915b33e8

File: ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/activationIntegrationTests.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    return expectSaga(
+      handleBonusActivationSaga,
+      bonusActivationSaga(startBonusActivation, getActivationById)
+    )
+      .withReducer(activationReducer)
+      .provide([
+        [
+          select(navigationCurrentRouteSelector),
+          fromNullable(BONUSVACANZE_ROUTES.ACTIVATION.LOADING)
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/activationIntegrationTests.ts]: changed in ab6915df4594f23fffa51ca6301c37b0ba322010

File: ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/activationIntegrationTests.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        ],
+        [
+          matchers.call.fn(startBonusActivation),
+          right({ status: 500, value: {} })
+        ],
+        [matchers.call.fn(getActivationById), right({ status: 500, value: {} })]
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/activationIntegrationTests.ts]: removed in ab6915df4594f23fffa51ca6301c37b0ba322010

File: ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/mockData.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    }
+  ],
+  expectedActions: [],
+  finalState: {
+    activation: { status: BonusActivationProgressEnum.ERROR },
+    allActive: {}
+  }
+} as MockBackendScenario;
+
+export const backendIntegrationTestCases = [
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/mockData.ts]: changed in ab6915df4594f23fffa51ca6301c37b0ba322010

File: ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/mockData.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  success,
+  eligibilityExpired,
+  bonusAlreadyExists,
+  error
+] as ReadonlyArray<MockBackendScenario>;
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/mockData.ts]: changed in ab6915df4594f23fffa51ca6301c37b0ba322010

File: ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/mockData.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  getBonusActivationResponseById: any;
+};
+
+interface MockBackendScenario extends IExpectedActions {
+  responses: ReadonlyArray<ActivationBackendResponse>;
+  finalState: MockActivationState;
+}
+
+// TODO: test polling timeout case
+export const success = {
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/mockData.ts]: changed in ab6915df4594f23fffa51ca6301c37b0ba322010 like the others, but I think for a const is the same thing, also the type assertion is a check for the compiler https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions

File: ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/activationIntegrationTests.ts (hunk 1 above)
Comment (fabriziofff) [reply ts/features/bonusVacanze/store/sagas/activation/__tests__/integration/activationIntegrationTests.ts:72]: I've changed the structure of `ActivationBackendResponse`, and now is possible to mock also the `left` scenario. I've added also the case you mentioned

``` typescript
const genericDecodingFailure = left([
  { context: [], value: new Error("decoding failure") }
]);
```

**Pull request merged**